pkg/frontend/v2/frontend_test.go

```diff
@@ -332,10 +341,59 @@ func TestFrontendFailedCancellation(t *testing.T) {
 	require.Nil(t, resp)
 
 	ms.checkWithLock(func() {
-		require.Equal(t, 1, len(ms.msgs))
+		require.Len(t, ms.msgs, 1)
 	})
 }
 
+// Test that the front-end shuts down gracefully, i.e. waits for in-flight queries to finish.
+func TestFrontend_GracefulShutdown(t *testing.T) {
+	const (
+		userID = "test"
+		body   = "all good"
+	)
+	f, _ := setupFrontend(t, nil, func(f *Frontend, msg *schedulerpb.FrontendToScheduler) *schedulerpb.SchedulerToFrontend {
+		// We cannot call QueryResult directly, as Frontend is not yet waiting for the response.
+		// It first needs to be told that enqueuing has succeeded.
+		go sendResponseWithDelay(t, f, 100*time.Millisecond, userID, msg.QueryID, &httpgrpc.HTTPResponse{
+			Code: 200,
+			Body: []byte(body),
+		})
+
+		return &schedulerpb.SchedulerToFrontend{Status: schedulerpb.OK}
+	})
+
+	resp, err := f.RoundTripGRPC(user.InjectOrgID(context.Background(), userID), &httpgrpc.HTTPRequest{})
+	require.NoError(t, err)
+	require.Equal(t, int32(200), resp.Code)
+	require.Equal(t, []byte(body), resp.Body)
+
+	require.NoError(t, services.StopAndAwaitTerminated(context.Background(), f))
+
+	wg := f.requests.wg.(*fakeWaitGroup)
```

Comment:
I'm wondering if we could really test the wait instead of using a mocked wait group. What if:
1. We change `sendResponseWithDelay(t, f, 100*time.Millisecond` to wait for a longer period (e.g. 2 seconds)
1. We run `f.RoundTripGRPC()` and its assertions in a dedicated goroutine
3. Then we call `services.StopAndAwaitTerminated(context.Background(), f)`, and we expect that the inflight query succeed (which means we have to wait for the goroutine who called `f.RoundTripGRPC()` to be done, because it asserts the actual query response

Then as a counterproof, you comment out your fix from `Frontend.stopping()` and ensure the test fails (the query should be canceled).

Reply:
I'll see if I can make this work. I was to be honest skeptical towards using timing in the test.

Reply:
@pracucci please see my rewritten test :) It catches when in-flight requests aren't waited on.